### ISSUE TITLE
fifelse followup

### ISSUE THREAD

**Issue by jangorecki** [Rdatatable/data.table#3740]
- names of `test` could be retained
```r
fifelse(c(a=T,b=F), 1:2, 2:1)
#[1] 1 1
ifelse(c(a=T,b=F), 1:2, 2:1)
#a b 
#1 1 
```
- recycling here needs to be fixed, AFAIU might result into segfault
```r
fifelse(c(TRUE, FALSE), 1:2, list(0L))
#[1]         1 544979712
ifelse(c(TRUE, FALSE), 1:2, list(0L))
#[[1]]
#[1] 1
#
#[[2]]
#[1] 0
```
- remove hutils, dplyr code from examples
- remove console output in examples

**Comment by jangorecki** [comment]
I would prefer `hutils`/`dplyr` behaviour to raise error when `yes` is a list and `no` is a atomic.
As a result user will need to wrap `no` into extra list
```r
DT = data.table(a = 0:5, b = list(1, list(2:3), list(4:6), list(6:4), list(3:2), 1))

DT[ , b := fifelse(a > 2, b, 0)]
# become
DT[ , b := fifelse(a > 2, b, list(0))][]
```
This shouldn't be big deal. I think it make sense when we want to use it on a `list` column.

**Comment by 2005m** [comment]
would you agree if `fifelse(c(TRUE, FALSE), 1:2, list(0L))` display an error like this :

```
Error in fifelse(c(TRUE, FALSE), 1:2, list(0L)) : 
  'yes' is of type integer but 'no' is of type list. Please make sure that both arguments have the same type.
```
and that we have the following ?

```
DT1 = data.table(a = 0:5, b = list(1, list(2:3), list(4:6), list(6:4), list(3:2), 1))
DT2 = data.table(a = 0:5, b = list(1, list(2:3), list(4:6), list(6:4), list(3:2), 1))
DT1 = DT1[ , b1 := fifelse(a > 2, b, 0)]
DT2 = DT2[ , b1 := fifelse(a > 2, b, list(0))]
identical(DT1,DT2)
[1] TRUE
```

Let me know if you want me to test something else. I don't want to commit anything until you are happy with the behaviour.

**Comment by jangorecki** [comment]
So for `len1 > 1` or `len2 > 1` branches raise an error.
I would personally avoid this kind of inconsistency, but lets wait for opinion of others. I also think wrapping in a `list` for values to meant to be a list column elements is not bad.

**Comment by 2005m** [comment]
Yes that is correct. This is consistent with what @MichaelChirico  ask me to implement in my pull request.

**Comment by shrektan** [comment]
Thanks for implementing this. `dplyr::if_else()` has a `missing` argument I think it's quite helpful, because the number of all the possible logic values in R are **3**, in fact:  `TRUE`, `FALSE`, `NA`. 

So do you think it's a good idea to add the `missing` argument in `fifelse()`?

Thanks.
